From: olh@suse.de
Subject: 2.6.11-rc2-bk3

 -- www.kernel.org/pub/linux/kernel/v2.6/snapshots/patch-2.6.11-rc2-bk2.log	2005-01-24 13:29:48.000000000 +0100
 ++ www.kernel.org/pub/linux/kernel/v2.6/snapshots/patch-2.6.11-rc2-bk3.log	2005-01-25 13:30:23.000000000 +0100
 @ -1,3 +1,62 @@
 ChangeSet@1.2020.1.8, 2005-01-24 17:41:45-08:00, davem@nuts.davemloft.net
   [TG3]: Update driver version and reldate.
   
   Signed-off-by: David S. Miller <davem@davemloft.net>
 
 ChangeSet@1.2020.1.7, 2005-01-24 17:39:51-08:00, mchan@broadcom.com
   [TG3]: Fix TSO for 5750
   
   - Fix TSO for 5750 chips by setting tcp checksum field to 0 for TSO packets
   - Add TG3_FLG2_HW_TSO flag for 5750 and newer chips that use the same TSO
     scheme
   
   Signed-off-by: Michael Chan <mchan@broadcom.com>
   Signed-off-by: David S. Miller <davem@davemloft.net>
 
 ChangeSet@1.2020.1.6, 2005-01-24 16:40:33-08:00, herbert@gondor.apana.org.au
   [IPV4/IPV6]: In ip_fragment(), reset ip_summed field on SKB sub-frags.
   
   If we forward a fragmented packet, we can have ip_summed
   set to CHECKSUM_HW or similar.  This is fine for local
   protocol processing, but once if we are forwarding this
   packet we want to reset ip_summed to CHECKSUM_NONE.
   
   Signed-off-by: Herbert Xu <herbert@gondor.apana.org.au>
   Signed-off-by: David S. Miller <davem@davemloft.net>
 
 ChangeSet@1.2027, 2005-01-24 15:57:20-08:00, arjan@infradead.org
   [PATCH] removing bcopy... because it's half broken
   
   Nothing in the kernel is using bcopy right know, and that is a good thing.
   Why? Because a lot of the architectures implement a broken bcopy()....
   the userspace standard bcopy() is basically a memmove() with a weird
   parameter order, however a bunch of architectures implement a memcpy() not a
   memmove().
   
   Instead of fixing this inconsistency, I decided to remove it entirely,
   explicit memcpy() and memmove() are prefered anyway (welcome to the 1990's)
   and nothing in the kernel is using these functions, so this saves code size
   as well for everyone.
   
   Signed-off-by: Arjan van de Ven <arjan@infradead.org>
   Signed-off-by: Adrian Bunk <bunk@stusta.de>
   
   [ Side note: the only reason for bcopy appears to be totally ancient
     gcc versions for OSF/1, used to originally cross-compile Linux on
     alpha. Possibly some other similar cases. Time to move on ;-]
   
   Signed-off-by: Linus Torvalds <torvalds@osdl.org>
 
 ChangeSet@1.2026, 2005-01-24 14:37:53-08:00, torvalds@ppc970.osdl.org
   Merge bk://linux-voyager.bkbits.net/voyager-2.6
   into ppc970.osdl.org:/home/torvalds/v2.6/linux
 
 ChangeSet@1.2025, 2005-01-24 14:36:59-08:00, torvalds@ppc970.osdl.org
   Free temporary pipe page after freeing the main buffers.
   
   Duh. Freeing the main buffers can _cause_ the temporary
   page to be created.
 
 ChangeSet@1.2020.1.5, 2005-01-23 11:09:59-08:00, torvalds@ppc970.osdl.org
   x86-64: don't crash and loop when the user passes an unknown earlyprintk= option.
   
diff -purN linux-2.6.11-rc2-bk2/Makefile linux-2.6.11-rc2-bk3/Makefile
--- linux-2.6.11-rc2-bk2/Makefile	2005-01-25 14:06:59.638323837 +0100
+++ linux-2.6.11-rc2-bk3/Makefile	2005-01-25 14:06:49.515435233 +0100
@@ -1,7 +1,7 @@
 VERSION = 2
 PATCHLEVEL = 6
 SUBLEVEL = 11
-EXTRAVERSION = -rc2-bk2
+EXTRAVERSION = -rc2-bk3
 NAME=Woozy Numbat
 
 # *DOCUMENTATION*
diff -purN linux-2.6.11-rc2-bk2/arch/alpha/lib/memmove.S linux-2.6.11-rc2-bk3/arch/alpha/lib/memmove.S
--- linux-2.6.11-rc2-bk2/arch/alpha/lib/memmove.S	2005-01-22 02:48:03.000000000 +0100
+++ linux-2.6.11-rc2-bk3/arch/alpha/lib/memmove.S	2005-01-25 14:06:49.516435078 +0100
@@ -12,18 +12,6 @@
 	.text
 
 	.align 4
-	.globl bcopy
-	.ent bcopy
-bcopy:
-	ldgp $29, 0($27)
-	.prologue 1
-	mov $16,$0
-	mov $17,$16
-	mov $0,$17
-	br  $31, memmove	!samegp
-	.end bcopy
-
-	.align 4
 	.globl memmove
 	.ent memmove
 memmove:
diff -purN linux-2.6.11-rc2-bk2/arch/i386/kernel/pci-dma.c linux-2.6.11-rc2-bk3/arch/i386/kernel/pci-dma.c
--- linux-2.6.11-rc2-bk2/arch/i386/kernel/pci-dma.c	2005-01-22 02:49:21.000000000 +0100
+++ linux-2.6.11-rc2-bk3/arch/i386/kernel/pci-dma.c	2005-01-25 14:06:49.524433833 +0100
@@ -122,6 +122,7 @@ void dma_release_declared_memory(struct 
 	if(!mem)
 		return;
 	dev->dma_mem = NULL;
+	iounmap(mem->virt_base);
 	kfree(mem->bitmap);
 	kfree(mem);
 }
diff -purN linux-2.6.11-rc2-bk2/arch/i386/kernel/reboot.c linux-2.6.11-rc2-bk3/arch/i386/kernel/reboot.c
--- linux-2.6.11-rc2-bk2/arch/i386/kernel/reboot.c	2005-01-22 02:47:32.000000000 +0100
+++ linux-2.6.11-rc2-bk3/arch/i386/kernel/reboot.c	2005-01-25 14:06:49.525433677 +0100
@@ -20,7 +20,7 @@
 void (*pm_power_off)(void);
 
 static int reboot_mode;
-int reboot_thru_bios;
+static int reboot_thru_bios;
 
 #ifdef CONFIG_SMP
 int reboot_smp = 0;
diff -purN linux-2.6.11-rc2-bk2/arch/i386/mach-visws/reboot.c linux-2.6.11-rc2-bk3/arch/i386/mach-visws/reboot.c
--- linux-2.6.11-rc2-bk2/arch/i386/mach-visws/reboot.c	2005-01-22 02:48:45.000000000 +0100
+++ linux-2.6.11-rc2-bk3/arch/i386/mach-visws/reboot.c	2005-01-25 14:06:49.526433522 +0100
@@ -8,9 +8,6 @@
 
 void (*pm_power_off)(void);
 
-int reboot_thru_bios;
-int reboot_smp;
-
 void machine_restart(char * __unused)
 {
 #ifdef CONFIG_SMP
diff -purN linux-2.6.11-rc2-bk2/arch/i386/mach-voyager/voyager_basic.c linux-2.6.11-rc2-bk3/arch/i386/mach-voyager/voyager_basic.c
--- linux-2.6.11-rc2-bk2/arch/i386/mach-voyager/voyager_basic.c	2005-01-22 02:47:31.000000000 +0100
+++ linux-2.6.11-rc2-bk3/arch/i386/mach-voyager/voyager_basic.c	2005-01-25 14:06:49.527433366 +0100
@@ -36,8 +36,6 @@
  */
 void (*pm_power_off)(void);
 
-int reboot_thru_bios;
-
 int voyager_level = 0;
 
 struct voyager_SUS *voyager_SUS = NULL;
diff -purN linux-2.6.11-rc2-bk2/arch/i386/mach-voyager/voyager_smp.c linux-2.6.11-rc2-bk3/arch/i386/mach-voyager/voyager_smp.c
--- linux-2.6.11-rc2-bk2/arch/i386/mach-voyager/voyager_smp.c	2005-01-22 02:47:48.000000000 +0100
+++ linux-2.6.11-rc2-bk3/arch/i386/mach-voyager/voyager_smp.c	2005-01-25 14:06:49.531432744 +0100
@@ -27,13 +27,10 @@
 #include <asm/mtrr.h>
 #include <asm/pgalloc.h>
 #include <asm/tlbflush.h>
-#include <asm/desc.h>
 #include <asm/arch_hooks.h>
 
 #include <linux/irq.h>
 
-int reboot_smp = 0;
-
 /* TLB state -- visible externally, indexed physically */
 DEFINE_PER_CPU(struct tlb_state, cpu_tlbstate) ____cacheline_aligned = { &init_mm, 0 };
 
diff -purN linux-2.6.11-rc2-bk2/arch/ia64/lib/memcpy.S linux-2.6.11-rc2-bk3/arch/ia64/lib/memcpy.S
--- linux-2.6.11-rc2-bk2/arch/ia64/lib/memcpy.S	2005-01-22 02:46:43.000000000 +0100
+++ linux-2.6.11-rc2-bk3/arch/ia64/lib/memcpy.S	2005-01-25 14:06:49.581424964 +0100
@@ -15,22 +15,6 @@
  */
 #include <asm/asmmacro.h>
 
-GLOBAL_ENTRY(bcopy)
-	.regstk 3,0,0,0
-	mov r8=in0
-	mov in0=in1
-	;;
-	mov in1=r8
-	// gas doesn't handle control flow across procedures, so it doesn't
-	// realize that a stop bit is needed before the "alloc" instruction
-	// below
-{
-	nop.m 0
-	nop.f 0
-	nop.i 0
-}	;;
-END(bcopy)
-	// FALL THROUGH
 GLOBAL_ENTRY(memcpy)
 
 #	define MEM_LAT	21		/* latency to memory */
diff -purN linux-2.6.11-rc2-bk2/arch/ia64/lib/memcpy_mck.S linux-2.6.11-rc2-bk3/arch/ia64/lib/memcpy_mck.S
--- linux-2.6.11-rc2-bk2/arch/ia64/lib/memcpy_mck.S	2005-01-22 02:49:22.000000000 +0100
+++ linux-2.6.11-rc2-bk3/arch/ia64/lib/memcpy_mck.S	2005-01-25 14:06:49.583424653 +0100
@@ -17,15 +17,6 @@
 
 #define EK(y...) EX(y)
 
-GLOBAL_ENTRY(bcopy)
-	.regstk 3,0,0,0
-	mov r8=in0
-	mov in0=in1
-	;;
-	mov in1=r8
-	;;
-END(bcopy)
-
 /* McKinley specific optimization */
 
 #define retval		r8
diff -purN linux-2.6.11-rc2-bk2/arch/parisc/lib/memcpy.c linux-2.6.11-rc2-bk3/arch/parisc/lib/memcpy.c
--- linux-2.6.11-rc2-bk2/arch/parisc/lib/memcpy.c	2005-01-22 02:49:12.000000000 +0100
+++ linux-2.6.11-rc2-bk3/arch/parisc/lib/memcpy.c	2005-01-25 14:06:49.604421385 +0100
@@ -515,16 +515,8 @@ void * memcpy(void * dst,const void *src
 	return dst;
 }
 
-void bcopy(const void * srcp, void * destp, size_t count)
-{
-	mtsp(get_kernel_space(), 1);
-	mtsp(get_kernel_space(), 2);
-	pa_memcpy(destp, srcp, count);
-}
-
 EXPORT_SYMBOL(copy_to_user);
 EXPORT_SYMBOL(copy_from_user);
 EXPORT_SYMBOL(copy_in_user);
 EXPORT_SYMBOL(memcpy);
-EXPORT_SYMBOL(bcopy);
 #endif
diff -purN linux-2.6.11-rc2-bk2/arch/ppc/boot/common/misc-common.c linux-2.6.11-rc2-bk3/arch/ppc/boot/common/misc-common.c
--- linux-2.6.11-rc2-bk2/arch/ppc/boot/common/misc-common.c	2005-01-22 02:48:01.000000000 +0100
+++ linux-2.6.11-rc2-bk3/arch/ppc/boot/common/misc-common.c	2005-01-25 14:06:49.605421230 +0100
@@ -52,7 +52,6 @@ extern char _end[];
 void puts(const char *);
 void putc(const char c);
 void puthex(unsigned long val);
-void _bcopy(char *src, char *dst, int len);
 void gunzip(void *, int, unsigned char *, int *);
 static int _cvt(unsigned long val, char *buf, long radix, char *digits);
 
diff -purN linux-2.6.11-rc2-bk2/arch/ppc/lib/string.S linux-2.6.11-rc2-bk3/arch/ppc/lib/string.S
--- linux-2.6.11-rc2-bk2/arch/ppc/lib/string.S	2005-01-22 02:48:26.000000000 +0100
+++ linux-2.6.11-rc2-bk3/arch/ppc/lib/string.S	2005-01-25 14:06:49.607420919 +0100
@@ -216,12 +216,6 @@ _GLOBAL(memset)
 	bdnz	8b
 	blr
 
-_GLOBAL(bcopy)
-	mr	r6,r3
-	mr	r3,r4
-	mr	r4,r6
-	b	memcpy
-
 /*
  * This version uses dcbz on the complete cache lines in the
  * destination area to reduce memory traffic.  This requires that
diff -purN linux-2.6.11-rc2-bk2/arch/ppc64/boot/string.S linux-2.6.11-rc2-bk3/arch/ppc64/boot/string.S
--- linux-2.6.11-rc2-bk2/arch/ppc64/boot/string.S	2005-01-22 02:46:42.000000000 +0100
+++ linux-2.6.11-rc2-bk3/arch/ppc64/boot/string.S	2005-01-25 14:06:49.608420763 +0100
@@ -96,13 +96,6 @@ memset:
 	bdnz	8b
 	blr
 
-	.globl	bcopy
-bcopy:
-	mr	r6,r3
-	mr	r3,r4
-	mr	r4,r6
-	b	memcpy
-
 	.globl	memmove
 memmove:
 	cmplw	0,r3,r4
diff -purN linux-2.6.11-rc2-bk2/arch/s390/lib/string.c linux-2.6.11-rc2-bk3/arch/s390/lib/string.c
--- linux-2.6.11-rc2-bk2/arch/s390/lib/string.c	2005-01-22 02:47:48.000000000 +0100
+++ linux-2.6.11-rc2-bk3/arch/s390/lib/string.c	2005-01-25 14:06:49.611420296 +0100
@@ -357,21 +357,6 @@ void *memcpy(void *dest, const void *src
 EXPORT_SYMBOL(memcpy);
 
 /**
- * bcopy - Copy one area of memory to another
- * @src: Where to copy from
- * @dest: Where to copy to
- * @n: The size of the area.
- *
- * Note that this is the same as memcpy(), with the arguments reversed.
- * memcpy() is the standard, bcopy() is a legacy BSD function.
- */
-void bcopy(const void *srcp, void *destp, size_t n)
-{
-	__builtin_memcpy(destp, srcp, n);
-}
-EXPORT_SYMBOL(bcopy);
-
-/**
  * memset - Fill a region of memory with the given value
  * @s: Pointer to the start of the area.
  * @c: The byte to fill the area with
diff -purN linux-2.6.11-rc2-bk2/arch/sparc/kernel/sparc_ksyms.c linux-2.6.11-rc2-bk3/arch/sparc/kernel/sparc_ksyms.c
--- linux-2.6.11-rc2-bk2/arch/sparc/kernel/sparc_ksyms.c	2005-01-22 02:47:00.000000000 +0100
+++ linux-2.6.11-rc2-bk3/arch/sparc/kernel/sparc_ksyms.c	2005-01-25 14:06:49.615419674 +0100
@@ -75,7 +75,6 @@ extern void *__memscan_generic(void *, i
 extern int __memcmp(const void *, const void *, __kernel_size_t);
 extern int __strncmp(const char *, const char *, __kernel_size_t);
 
-extern void bcopy (const char *, char *, int);
 extern int __ashrdi3(int, int);
 extern int __ashldi3(int, int);
 extern int __lshrdi3(int, int);
@@ -261,7 +260,6 @@ EXPORT_SYMBOL(__prom_getchild);
 EXPORT_SYMBOL(__prom_getsibling);
 
 /* sparc library symbols */
-EXPORT_SYMBOL(bcopy);
 EXPORT_SYMBOL(memchr);
 EXPORT_SYMBOL(memscan);
 EXPORT_SYMBOL(strlen);
diff -purN linux-2.6.11-rc2-bk2/arch/sparc/lib/memcpy.S linux-2.6.11-rc2-bk3/arch/sparc/lib/memcpy.S
--- linux-2.6.11-rc2-bk2/arch/sparc/lib/memcpy.S	2005-01-22 02:49:19.000000000 +0100
+++ linux-2.6.11-rc2-bk3/arch/sparc/lib/memcpy.S	2005-01-25 14:06:49.617419363 +0100
@@ -1,5 +1,5 @@
-/* memcpy.S: Sparc optimized memcpy, bcopy and memmove code
- * Hand optimized from GNU libc's memcpy, bcopy and memmove
+/* memcpy.S: Sparc optimized memcpy and memmove code
+ * Hand optimized from GNU libc's memcpy and memmove
  * Copyright (C) 1991,1996 Free Software Foundation
  * Copyright (C) 1995 Linus Torvalds (Linus.Torvalds@helsinki.fi)
  * Copyright (C) 1996 David S. Miller (davem@caip.rutgers.edu)
@@ -192,13 +192,6 @@ x:
 	retl
 	 nop		! Only bcopy returns here and it retuns void...
 
-FUNC(bcopy)
-	mov		%o0, %o3
-	mov		%o1, %o0
-	mov		%o3, %o1
-	tst		%o2
-	bcs		0b
-	 /* Do the cmp in the delay slot */
 #ifdef __KERNEL__
 FUNC(amemmove)
 FUNC(__memmove)
diff -purN linux-2.6.11-rc2-bk2/arch/v850/lib/memcpy.c linux-2.6.11-rc2-bk3/arch/v850/lib/memcpy.c
--- linux-2.6.11-rc2-bk2/arch/v850/lib/memcpy.c	2005-01-22 02:47:15.000000000 +0100
+++ linux-2.6.11-rc2-bk3/arch/v850/lib/memcpy.c	2005-01-25 14:06:49.623418429 +0100
@@ -60,11 +60,6 @@ void *memcpy (void *dst, const void *src
 	return dst;
 }
 
-void bcopy (const char *src, char *dst, int size)
-{
-	memcpy (dst, src, size);
-}
-
 void *memmove (void *dst, const void *src, __kernel_size_t size)
 {
 	if ((unsigned long)dst < (unsigned long)src
diff -purN linux-2.6.11-rc2-bk2/drivers/net/tg3.c linux-2.6.11-rc2-bk3/drivers/net/tg3.c
--- linux-2.6.11-rc2-bk2/drivers/net/tg3.c	2005-01-25 14:06:59.940414147 +0100
+++ linux-2.6.11-rc2-bk3/drivers/net/tg3.c	2005-01-25 14:06:49.846383732 +0100
@@ -60,8 +60,8 @@
 
 #define DRV_MODULE_NAME		"tg3"
 #define PFX DRV_MODULE_NAME	": "
-#define DRV_MODULE_VERSION	"3.17"
-#define DRV_MODULE_RELDATE	"January 22, 2005"
+#define DRV_MODULE_VERSION	"3.18"
+#define DRV_MODULE_RELDATE	"January 24, 2005"
 
 #define TG3_DEF_MAC_MODE	0
 #define TG3_DEF_RX_MODE		0
@@ -3111,11 +3111,19 @@ static int tg3_start_xmit(struct sk_buff
 
 		skb->nh.iph->check = 0;
 		skb->nh.iph->tot_len = ntohs(mss + ip_tcp_len + tcp_opt_len);
-		skb->h.th->check = ~csum_tcpudp_magic(skb->nh.iph->saddr,
-						      skb->nh.iph->daddr,
-						      0, IPPROTO_TCP, 0);
+		if (tp->tg3_flags2 & TG3_FLG2_HW_TSO) {
+			skb->h.th->check = 0;
+			base_flags &= ~TXD_FLAG_TCPUDP_CSUM;
+		}
+		else {
+			skb->h.th->check =
+				~csum_tcpudp_magic(skb->nh.iph->saddr,
+						   skb->nh.iph->daddr,
+						   0, IPPROTO_TCP, 0);
+		}
 
-		if (GET_ASIC_REV(tp->pci_chip_rev_id) == ASIC_REV_5705) {
+		if ((tp->tg3_flags2 & TG3_FLG2_HW_TSO) ||
+		    (GET_ASIC_REV(tp->pci_chip_rev_id) == ASIC_REV_5705)) {
 			if (tcp_opt_len || skb->nh.iph->ihl > 5) {
 				int tsflags;
 
@@ -3182,7 +3190,7 @@ static int tg3_start_xmit(struct sk_buff
 				would_hit_hwbug = entry + 1;
 			}
 
-			if (GET_ASIC_REV(tp->pci_chip_rev_id) == ASIC_REV_5750)
+			if (tp->tg3_flags2 & TG3_FLG2_HW_TSO)
 				tg3_set_txd(tp, entry, mapping, len,
 					    base_flags, (i == last)|(mss << 1));
 			else
@@ -4774,7 +4782,7 @@ static int tg3_load_tso_firmware(struct 
 	unsigned long cpu_base, cpu_scratch_base, cpu_scratch_size;
 	int err, i;
 
-	if (GET_ASIC_REV(tp->pci_chip_rev_id) == ASIC_REV_5750)
+	if (tp->tg3_flags2 & TG3_FLG2_HW_TSO)
 		return 0;
 
 	if (GET_ASIC_REV(tp->pci_chip_rev_id) == ASIC_REV_5705) {
@@ -5208,7 +5216,7 @@ static int tg3_reset_hw(struct tg3 *tp)
 	}
 
 #if TG3_TSO_SUPPORT != 0
-	if (GET_ASIC_REV(tp->pci_chip_rev_id) == ASIC_REV_5750)
+	if (tp->tg3_flags2 & TG3_FLG2_HW_TSO)
 		rdmac_mode |= (1 << 27);
 #endif
 
@@ -5358,7 +5366,7 @@ static int tg3_reset_hw(struct tg3 *tp)
 	tw32(RCVDBDI_MODE, RCVDBDI_MODE_ENABLE | RCVDBDI_MODE_INV_RING_SZ);
 	tw32(SNDDATAI_MODE, SNDDATAI_MODE_ENABLE);
 #if TG3_TSO_SUPPORT != 0
-	if (GET_ASIC_REV(tp->pci_chip_rev_id) == ASIC_REV_5750)
+	if (tp->tg3_flags2 & TG3_FLG2_HW_TSO)
 		tw32(SNDDATAI_MODE, SNDDATAI_MODE_ENABLE | 0x8);
 #endif
 	tw32(SNDBDI_MODE, SNDBDI_MODE_ENABLE | SNDBDI_MODE_ATTN_ENABLE);
@@ -7867,6 +7875,9 @@ static int __devinit tg3_get_invariants(
 	tp->pci_hdr_type     = (cacheline_sz_reg >> 16) & 0xff;
 	tp->pci_bist         = (cacheline_sz_reg >> 24) & 0xff;
 
+	if (GET_ASIC_REV(tp->pci_chip_rev_id) == ASIC_REV_5750)
+		tp->tg3_flags2 |= TG3_FLG2_HW_TSO;
+
 	if (pci_find_capability(tp->pdev, PCI_CAP_ID_EXP) != 0)
 		tp->tg3_flags2 |= TG3_FLG2_PCI_EXPRESS;
 
@@ -8762,11 +8773,13 @@ static int __devinit tg3_init_one(struct
 	}
 
 #if TG3_TSO_SUPPORT != 0
-	if (GET_ASIC_REV(tp->pci_chip_rev_id) == ASIC_REV_5700 ||
+	if (tp->tg3_flags2 & TG3_FLG2_HW_TSO) {
+		tp->tg3_flags2 |= TG3_FLG2_TSO_CAPABLE;
+	}
+	else if (GET_ASIC_REV(tp->pci_chip_rev_id) == ASIC_REV_5700 ||
 	    GET_ASIC_REV(tp->pci_chip_rev_id) == ASIC_REV_5701 ||
 	    tp->pci_chip_rev_id == CHIPREV_ID_5705_A0 ||
-	    ((tp->tg3_flags & TG3_FLAG_ENABLE_ASF) != 0 &&
-	     GET_ASIC_REV(tp->pci_chip_rev_id) != ASIC_REV_5750)) {
+	    (tp->tg3_flags & TG3_FLAG_ENABLE_ASF) != 0) {
 		tp->tg3_flags2 &= ~TG3_FLG2_TSO_CAPABLE;
 	} else {
 		tp->tg3_flags2 |= TG3_FLG2_TSO_CAPABLE;
diff -purN linux-2.6.11-rc2-bk2/drivers/net/tg3.h linux-2.6.11-rc2-bk3/drivers/net/tg3.h
--- linux-2.6.11-rc2-bk2/drivers/net/tg3.h	2005-01-25 14:06:59.946413213 +0100
+++ linux-2.6.11-rc2-bk3/drivers/net/tg3.h	2005-01-25 14:06:49.852382798 +0100
@@ -2105,6 +2105,7 @@ struct tg3 {
 #define TG3_FLG2_PHY_SERDES		0x00002000
 #define TG3_FLG2_CAPACITIVE_COUPLING	0x00004000
 #define TG3_FLG2_FLASH			0x00008000
+#define TG3_FLG2_HW_TSO			0x00010000
 
 	u32				split_mode_max_reqs;
 #define SPLIT_MODE_5704_MAX_REQ		3
diff -purN linux-2.6.11-rc2-bk2/fs/pipe.c linux-2.6.11-rc2-bk3/fs/pipe.c
--- linux-2.6.11-rc2-bk2/fs/pipe.c	2005-01-22 02:48:19.000000000 +0100
+++ linux-2.6.11-rc2-bk3/fs/pipe.c	2005-01-25 14:06:49.867380465 +0100
@@ -630,13 +630,13 @@ void free_pipe_info(struct inode *inode)
 	struct pipe_inode_info *info = inode->i_pipe;
 
 	inode->i_pipe = NULL;
-	if (info->tmp_page)
-		__free_page(info->tmp_page);
 	for (i = 0; i < PIPE_BUFFERS; i++) {
 		struct pipe_buffer *buf = info->bufs + i;
 		if (buf->ops)
 			buf->ops->release(info, buf);
 	}
+	if (info->tmp_page)
+		__free_page(info->tmp_page);
 	kfree(info);
 }
 
diff -purN linux-2.6.11-rc2-bk2/include/asm-alpha/string.h linux-2.6.11-rc2-bk3/include/asm-alpha/string.h
--- linux-2.6.11-rc2-bk2/include/asm-alpha/string.h	2005-01-22 02:47:48.000000000 +0100
+++ linux-2.6.11-rc2-bk3/include/asm-alpha/string.h	2005-01-25 14:06:49.868380309 +0100
@@ -13,7 +13,6 @@
 #define __HAVE_ARCH_MEMCPY
 extern void * memcpy(void *, const void *, size_t);
 #define __HAVE_ARCH_MEMMOVE
-#define __HAVE_ARCH_BCOPY
 extern void * memmove(void *, const void *, size_t);
 
 /* For backward compatibility with modules.  Unused otherwise.  */
diff -purN linux-2.6.11-rc2-bk2/include/asm-arm/string.h linux-2.6.11-rc2-bk3/include/asm-arm/string.h
--- linux-2.6.11-rc2-bk2/include/asm-arm/string.h	2005-01-22 02:48:17.000000000 +0100
+++ linux-2.6.11-rc2-bk3/include/asm-arm/string.h	2005-01-25 14:06:49.891376730 +0100
@@ -25,8 +25,6 @@ extern void * memchr(const void *, int, 
 #define __HAVE_ARCH_MEMSET
 extern void * memset(void *, int, __kernel_size_t);
 
-#define __HAVE_ARCH_BCOPY
-
 extern void __memzero(void *ptr, __kernel_size_t n);
 
 #define memset(p,v,n)							\
diff -purN linux-2.6.11-rc2-bk2/include/asm-ia64/string.h linux-2.6.11-rc2-bk3/include/asm-ia64/string.h
--- linux-2.6.11-rc2-bk2/include/asm-ia64/string.h	2005-01-22 02:48:48.000000000 +0100
+++ linux-2.6.11-rc2-bk3/include/asm-ia64/string.h	2005-01-25 14:06:49.904374708 +0100
@@ -14,7 +14,6 @@
 #define __HAVE_ARCH_STRLEN	1 /* see arch/ia64/lib/strlen.S */
 #define __HAVE_ARCH_MEMSET	1 /* see arch/ia64/lib/memset.S */
 #define __HAVE_ARCH_MEMCPY	1 /* see arch/ia64/lib/memcpy.S */
-#define __HAVE_ARCH_BCOPY	1 /* see arch/ia64/lib/memcpy.S */
 
 extern __kernel_size_t strlen (const char *);
 extern void *memcpy (void *, const void *, __kernel_size_t);
diff -purN linux-2.6.11-rc2-bk2/include/asm-mips/string.h linux-2.6.11-rc2-bk3/include/asm-mips/string.h
--- linux-2.6.11-rc2-bk2/include/asm-mips/string.h	2005-01-22 02:46:41.000000000 +0100
+++ linux-2.6.11-rc2-bk3/include/asm-mips/string.h	2005-01-25 14:06:49.908374085 +0100
@@ -137,9 +137,6 @@ extern void *memcpy(void *__to, __const_
 #define __HAVE_ARCH_MEMMOVE
 extern void *memmove(void *__dest, __const__ void *__src, size_t __n);
 
-/* Don't build bcopy at all ...  */
-#define __HAVE_ARCH_BCOPY
-
 #ifdef CONFIG_MIPS32
 #define __HAVE_ARCH_MEMSCAN
 static __inline__ void *memscan(void *__addr, int __c, size_t __size)
diff -purN linux-2.6.11-rc2-bk2/include/asm-parisc/string.h linux-2.6.11-rc2-bk3/include/asm-parisc/string.h
--- linux-2.6.11-rc2-bk2/include/asm-parisc/string.h	2005-01-22 02:48:28.000000000 +0100
+++ linux-2.6.11-rc2-bk3/include/asm-parisc/string.h	2005-01-25 14:06:49.908374085 +0100
@@ -7,7 +7,4 @@ extern void * memset(void *, int, size_t
 #define __HAVE_ARCH_MEMCPY
 void * memcpy(void * dest,const void *src,size_t count);
 
-#define __HAVE_ARCH_BCOPY
-void bcopy(const void * srcp, void * destp, size_t count);
-
 #endif
diff -purN linux-2.6.11-rc2-bk2/include/asm-ppc/string.h linux-2.6.11-rc2-bk3/include/asm-ppc/string.h
--- linux-2.6.11-rc2-bk2/include/asm-ppc/string.h	2005-01-22 02:48:35.000000000 +0100
+++ linux-2.6.11-rc2-bk3/include/asm-ppc/string.h	2005-01-25 14:06:49.909373930 +0100
@@ -9,7 +9,6 @@
 #define __HAVE_ARCH_STRCMP
 #define __HAVE_ARCH_STRCAT
 #define __HAVE_ARCH_MEMSET
-#define __HAVE_ARCH_BCOPY
 #define __HAVE_ARCH_MEMCPY
 #define __HAVE_ARCH_MEMMOVE
 #define __HAVE_ARCH_MEMCMP
diff -purN linux-2.6.11-rc2-bk2/include/asm-ppc64/string.h linux-2.6.11-rc2-bk3/include/asm-ppc64/string.h
--- linux-2.6.11-rc2-bk2/include/asm-ppc64/string.h	2005-01-22 02:48:28.000000000 +0100
+++ linux-2.6.11-rc2-bk3/include/asm-ppc64/string.h	2005-01-25 14:06:49.910373774 +0100
@@ -14,7 +14,6 @@
 #define __HAVE_ARCH_STRCMP
 #define __HAVE_ARCH_STRCAT
 #define __HAVE_ARCH_MEMSET
-#define __HAVE_ARCH_BCOPY
 #define __HAVE_ARCH_MEMCPY
 #define __HAVE_ARCH_MEMMOVE
 #define __HAVE_ARCH_MEMCMP
diff -purN linux-2.6.11-rc2-bk2/include/asm-s390/string.h linux-2.6.11-rc2-bk3/include/asm-s390/string.h
--- linux-2.6.11-rc2-bk2/include/asm-s390/string.h	2005-01-22 02:49:19.000000000 +0100
+++ linux-2.6.11-rc2-bk3/include/asm-s390/string.h	2005-01-25 14:06:49.910373774 +0100
@@ -15,7 +15,6 @@
 #include <linux/types.h>
 #endif
 
-#define __HAVE_ARCH_BCOPY	/* arch function */
 #define __HAVE_ARCH_MEMCHR	/* inline & arch function */
 #define __HAVE_ARCH_MEMCMP	/* arch function */
 #define __HAVE_ARCH_MEMCPY	/* gcc builtin & arch function */
diff -purN linux-2.6.11-rc2-bk2/include/asm-sh/string.h linux-2.6.11-rc2-bk3/include/asm-sh/string.h
--- linux-2.6.11-rc2-bk2/include/asm-sh/string.h	2005-01-22 02:48:48.000000000 +0100
+++ linux-2.6.11-rc2-bk3/include/asm-sh/string.h	2005-01-25 14:06:49.911373618 +0100
@@ -124,7 +124,4 @@ extern void *memchr(const void *__s, int
 #define __HAVE_ARCH_STRLEN
 extern size_t strlen(const char *);
 
-/* Don't build bcopy at all ...  */
-#define __HAVE_ARCH_BCOPY
-
 #endif /* __ASM_SH_STRING_H */
diff -purN linux-2.6.11-rc2-bk2/include/asm-sparc/string.h linux-2.6.11-rc2-bk3/include/asm-sparc/string.h
--- linux-2.6.11-rc2-bk2/include/asm-sparc/string.h	2005-01-22 02:46:43.000000000 +0100
+++ linux-2.6.11-rc2-bk3/include/asm-sparc/string.h	2005-01-25 14:06:49.912373463 +0100
@@ -22,7 +22,6 @@ extern __kernel_size_t __memset(void *,i
 #ifndef EXPORT_SYMTAB_STROPS
 
 /* First the mem*() things. */
-#define __HAVE_ARCH_BCOPY
 #define __HAVE_ARCH_MEMMOVE
 #undef memmove
 #define memmove(_to, _from, _n) \
diff -purN linux-2.6.11-rc2-bk2/include/asm-v850/string.h linux-2.6.11-rc2-bk3/include/asm-v850/string.h
--- linux-2.6.11-rc2-bk2/include/asm-v850/string.h	2005-01-22 02:49:21.000000000 +0100
+++ linux-2.6.11-rc2-bk3/include/asm-v850/string.h	2005-01-25 14:06:49.913373307 +0100
@@ -14,13 +14,11 @@
 #ifndef __V850_STRING_H__
 #define __V850_STRING_H__
 
-#define __HAVE_ARCH_BCOPY
 #define __HAVE_ARCH_MEMCPY
 #define __HAVE_ARCH_MEMSET
 #define __HAVE_ARCH_MEMMOVE
 
 extern void *memcpy (void *, const void *, __kernel_size_t);
-extern void bcopy (const char *, char *, int);
 extern void *memset (void *, int, __kernel_size_t);
 extern void *memmove (void *, const void *, __kernel_size_t);
 
diff -purN linux-2.6.11-rc2-bk2/lib/string.c linux-2.6.11-rc2-bk3/lib/string.c
--- linux-2.6.11-rc2-bk2/lib/string.c	2005-01-22 02:48:48.000000000 +0100
+++ linux-2.6.11-rc2-bk3/lib/string.c	2005-01-25 14:06:49.922371907 +0100
@@ -454,30 +454,6 @@ void * memset(void * s,int c,size_t coun
 EXPORT_SYMBOL(memset);
 #endif
 
-#ifndef __HAVE_ARCH_BCOPY
-/**
- * bcopy - Copy one area of memory to another
- * @srcp: Where to copy from
- * @destp: Where to copy to
- * @count: The size of the area.
- *
- * Note that this is the same as memcpy(), with the arguments reversed.
- * memcpy() is the standard, bcopy() is a legacy BSD function.
- *
- * You should not use this function to access IO space, use memcpy_toio()
- * or memcpy_fromio() instead.
- */
-void bcopy(const void * srcp, void * destp, size_t count)
-{
-	const char *src = srcp;
-	char *dest = destp;
-
-	while (count--)
-		*dest++ = *src++;
-}
-EXPORT_SYMBOL(bcopy);
-#endif
-
 #ifndef __HAVE_ARCH_MEMCPY
 /**
  * memcpy - Copy one area of memory to another
diff -purN linux-2.6.11-rc2-bk2/net/ipv4/ip_output.c linux-2.6.11-rc2-bk3/net/ipv4/ip_output.c
--- linux-2.6.11-rc2-bk2/net/ipv4/ip_output.c	2005-01-22 02:49:19.000000000 +0100
+++ linux-2.6.11-rc2-bk3/net/ipv4/ip_output.c	2005-01-25 14:06:49.944368484 +0100
@@ -504,6 +504,7 @@ int ip_fragment(struct sk_buff *skb, int
 			/* Prepare header of the next frame,
 			 * before previous one went down. */
 			if (frag) {
+				frag->ip_summed = CHECKSUM_NONE;
 				frag->h.raw = frag->data;
 				frag->nh.raw = __skb_push(frag, hlen);
 				memcpy(frag->nh.raw, iph, hlen);
diff -purN linux-2.6.11-rc2-bk2/net/ipv6/ip6_output.c linux-2.6.11-rc2-bk3/net/ipv6/ip6_output.c
--- linux-2.6.11-rc2-bk2/net/ipv6/ip6_output.c	2005-01-22 02:46:59.000000000 +0100
+++ linux-2.6.11-rc2-bk3/net/ipv6/ip6_output.c	2005-01-25 14:06:49.946368173 +0100
@@ -592,6 +592,7 @@ static int ip6_fragment(struct sk_buff *
 			/* Prepare header of the next frame,
 			 * before previous one went down. */
 			if (frag) {
+				frag->ip_summed = CHECKSUM_NONE;
 				frag->h.raw = frag->data;
 				fh = (struct frag_hdr*)__skb_push(frag, sizeof(struct frag_hdr));
 				frag->nh.raw = __skb_push(frag, hlen);
